packages/fx-core/src/core/middleware/utils/v3MigrationUtils.ts

```diff
@@ -80,3 +85,50 @@ export async function getProjectVersion(ctx: CoreHookContext): Promise<string> {
   }
   return "0.0.0";
 }
+
+export function parseCapabilities(projectSettings: ProjectSettings): {
+  TabSso: boolean;
+  BotSso: boolean;
+} {
+  let tabSso, botSso;
+  if ((projectSettings as ProjectSettingsV3).components) {
+    tabSso = (projectSettings as ProjectSettingsV3).components.some((component, index, obj) => {
+      return component.name === "teams-tab" && component.sso == true;
+    });
+    botSso = (projectSettings as ProjectSettingsV3).components.some((component, index, obj) => {
+      return component.name === "teams-bot" && component.sso == true;
+    });
+  } else {
+    // For projects that does not componentize.
+    // Should not be called.
```

Comment:
what does this mean?

Reply:
This part is for scenario that projects migrate from v2 - not componentized. I am not quite sure whether this scenario is valid, but from previous discussion, we don't expect this scenario.